lib/to-lua.nix

```diff
@@ -32,51 +32,184 @@ rec {
+          value
+        else if allowRawValues && value ? __raw then
+          value
+        else if isList value then
+          map removeEmptiesRecursive value
+        else if isAttrs value then
+          concatMapAttrs (
             n: v:
-            let
-              keyString =
-                if n == "__emptyString" then
-                  "['']"
-                else if hasPrefix "__rawKey__" n then
-                  "[${removePrefix "__rawKey__" n}]"
-                else if isIdentifier n then
-                  n
-                else
-                  "[${toLuaObject n}]";
-              valueString = toLuaObject v;
-            in
-            if hasPrefix "__unkeyed" n then valueString else "${keyString} = ${valueString}"
-          ) (filterAttrs (n: v: v != null && (toLuaObject v != "{}")) args)
-        ))
-        + "}"
-    else if builtins.isList args then
-      "{" + concatMapStringsSep ", " toLuaObject args + "}"
-    else if builtins.isString args then
-      # This should be enough!
-      builtins.toJSON args
-    else if builtins.isPath args then
-      builtins.toJSON (toString args)
-    else if builtins.isBool args then
-      "${boolToString args}"
-    else if builtins.isFloat args then
-      "${toString args}"
-    else if builtins.isInt args then
-      "${toString args}"
-    else if (args == null) then
-      "nil"
-    else
-      "";
+            if removeNullAttrValues && v == null then
+              { }
+            else if removeEmptyAttrValues && (v == [ ] || v == { }) then
+              { }
+            else if allowExplicitEmpty && v ? __empty then
+              { ${n} = { }; }
+            else if isAttrs v then
+              let
+                v' = removeEmptiesRecursive v;
+              in
+              if v' == { } then { } else { ${n} = v'; }
+            else
+              { ${n} = v; }
+          ) value
+        else
+          value;
+
+      toTableKey =
+        s:
+        if allowRawAttrKeys && hasPrefix "__rawKey__" s then
+          "[${removePrefix "__rawKey__" s}]"
+        else if allowUnquotedAttrKeys && isIdentifier s then
+          s
+        else if allowLegacyEmptyStringAttr && s == "__emptyString" then
+          trace ''nixvim(toLua): __emptyString is deprecated, just use an attribute named "".'' (
+            toTableKey ""
+          )
+        else
+          "[${go "" s}]";
+
+      go =
+        indent: v:
+        let
+          introSpace = if multiline then "\n${indent}  " else " ";
+          outroSpace = if multiline then "\n${indent}" else " ";
+        in
+        if v == null then
+          "nil"
+        else if isInt v then
+          toString v
+        # toString loses precision on floats, so we use toJSON instead. This isn't perfect
+        # as the resulting string may not parse back as a float (e.g. 42, 1e-06), but for
+        # pretty-printing purposes this is acceptable.
+        else if isFloat v then
+          builtins.toJSON v
```

Comment:
IDK what the best approach is for lua. Previously we just did `toString v`, but the precision notice copied from `toPretty` is kinda scary :confused:

Reply:
Not a problem for lua I think, the only data type is "number"

Reply:
Ok, I'll amend the comment to something along the lines of "toString can reduce precision, so we use toJSON. Although this can convert to int, that shouldn't matter for lua."

Reply:
I've done some experimenting in the nix and lua repls, it seems toJSON is indeed better and lua seems to understand the exponetised floats, although I can't find any reference to that in the [5.1 ref manual](https://www.lua.org/manual/5.1/manual.html).

```
nix-repl> builtins.toJSON 400034582305840239.3
"4.0003458230584026e+17"

lua> print (400034582305840239.3)
4.0003458230584e+17

lua> print (4.0003458230584026e+17)
4.0003458230584e+17

nix-repl> builtins.toJSON (4 * 0.000001)
"4e-06"

lua> print (4e-06)
4e-06

nix-repl> builtins.toJSON (4 * 0.000001 + 0.01)
"0.010004"

lua> print (4e-06 + 0.01)
0.010004
```

The comment from `toPretty` seems to be referencing how _nix_ can't parse exponentiatized float literals:

```
nix-repl> builtins.toJSON (4e-06)
error: undefined variable 'e-06'
       at «string»:1:19:
            1| builtins.toJSON (4e-06)
             |                   ^
```